File: ios/LivenessDetection/PoseEstimator.swift
import Foundation
import MediaPipeTasksVision

struct Pose {
  let yaw: Float
  let pitch: Float
  let roll: Float
}

enum PoseEstimator {
  static func fromTransformationMatrix(_ matrix: [Float]?) -> Pose? {
    guard let matrix = matrix, matrix.count >= 16 else { return nil }
    let r00 = matrix[0]
    let r10 = matrix[4]
    let r20 = matrix[8]
    let r21 = matrix[9]
    let r22 = matrix[10]

    let pitch = asin(-r20)
    let yaw = atan2(r10, r00)
    let roll = atan2(r21, r22)

    return Pose(
      yaw: radToDeg(yaw),
      pitch: radToDeg(pitch),
      roll: radToDeg(roll)
    )
  }

  static func fromLandmarks(_ landmarks: [NormalizedLandmark]) -> Pose {
    let leftEyeOuter = landmarks[33]
    let rightEyeOuter = landmarks[263]
    let noseTip = landmarks[1]
    let chin = landmarks[152]

    let yaw = atan2(
      rightEyeOuter.z - leftEyeOuter.z,
      rightEyeOuter.x - leftEyeOuter.x
    )
    let roll = atan2(
      rightEyeOuter.y - leftEyeOuter.y,
      rightEyeOuter.x - leftEyeOuter.x
    )
    let pitch = atan2(
      chin.y - noseTip.y,
      chin.z - noseTip.z
    )

    return Pose(
      yaw: radToDeg(yaw),
      pitch: radToDeg(pitch),
      roll: radToDeg(roll)
    )
  }

  private static func radToDeg(_ value: Float) -> Float {
    return value * 180 / Float.pi
  }
}

File: android/liveness/src/main/java/com/liveness/detection/PoseEstimator.kt
package com.liveness.detection

import com.google.mediapipe.tasks.components.containers.NormalizedLandmark
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.PI

data class Pose(val yaw: Float, val pitch: Float, val roll: Float)

object PoseEstimator {
  fun fromTransformationMatrix(matrixData: FloatArray?): Pose? {
    if (matrixData == null || matrixData.size < 16) return null

    val r00 = matrixData[0]
    val r01 = matrixData[1]
    val r02 = matrixData[2]
    val r10 = matrixData[4]
    val r11 = matrixData[5]
    val r12 = matrixData[6]
    val r20 = matrixData[8]
    val r21 = matrixData[9]
    val r22 = matrixData[10]

    val pitch = asin(-r20)
    val yaw = atan2(r10, r00)
    val roll = atan2(r21, r22)

    return Pose(
      yaw = radToDeg(yaw),
      pitch = radToDeg(pitch),
      roll = radToDeg(roll),
    )
  }

  fun fromLandmarks(landmarks: List<NormalizedLandmark>): Pose {
    val leftEyeOuter = landmarks[33]
    val rightEyeOuter = landmarks[263]
    val noseTip = landmarks[1]
    val chin = landmarks[152]

    val yaw = atan2(
      rightEyeOuter.z() - leftEyeOuter.z(),
      rightEyeOuter.x() - leftEyeOuter.x(),
    )
    val roll = atan2(
      rightEyeOuter.y() - leftEyeOuter.y(),
      rightEyeOuter.x() - leftEyeOuter.x(),
    )
    val pitch = atan2(
      chin.y() - noseTip.y(),
      chin.z() - noseTip.z(),
    )

    return Pose(
      yaw = radToDeg(yaw),
      pitch = radToDeg(pitch),
      roll = radToDeg(roll),
    )
  }

  private fun radToDeg(value: Float): Float {
    return (value * 180f / PI.toFloat())
  }
}
